File: python/oauth/copilot/oauth.py
"""
GitHub Copilot OAuth device flow implementation.

Implements the OAuth2 device authorization flow for GitHub Copilot.
"""

from dataclasses import dataclass
from typing import Optional
import json
import time
import requests
from urllib.parse import urlencode

from ..token import Token
from .http import get_headers


# GitHub OAuth endpoints
CLIENT_ID = "Iv1.b507a08c87ecfe98"
DEVICE_CODE_URL = "https://github.com/login/device/code"
ACCESS_TOKEN_URL = "https://github.com/login/oauth/access_token"
COPILOT_TOKEN_URL = "https://api.github.com/copilot_internal/v2/token"

# Maximum response size (1MB) to prevent memory exhaustion
MAX_RESPONSE_SIZE = 1024 * 1024

# Custom exceptions
class NotAvailableError(Exception):
    """Raised when GitHub Copilot is not available for the user."""
    pass


@dataclass
class DeviceCode:
    """
    Device code response from GitHub.
    
    Attributes:
        device_code: Code used for polling
        user_code: Code displayed to the user
        verification_uri: URL where user authorizes the device
        expires_in: Seconds until the device code expires
        interval: Minimum polling interval in seconds
    """
    device_code: str
    user_code: str
    verification_uri: str
    expires_in: int
    interval: int

    @classmethod
    def from_dict(cls, data: dict) -> "DeviceCode":
        """Create DeviceCode from API response."""
        return cls(
            device_code=data["device_code"],
            user_code=data["user_code"],
            verification_uri=data["verification_uri"],
            expires_in=data["expires_in"],
            interval=data.get("interval", 5),
        )


def request_device_code(timeout: int = 30) -> DeviceCode:
    """
    Initiate the device code flow with GitHub.
    
    Args:
        timeout: Request timeout in seconds
        
    Returns:
        DeviceCode containing user_code and verification_uri
        
    Raises:
        requests.RequestException: If the request fails
    """
    data = {
        "client_id": CLIENT_ID,
        "scope": "read:user",
    }
    
    headers = {
        "Accept": "application/json",
        "Content-Type": "application/x-www-form-urlencoded",
        "User-Agent": get_headers()["User-Agent"],
    }
    
    response = requests.post(
        DEVICE_CODE_URL,
        data=urlencode(data),
        headers=headers,
        timeout=timeout,
    )
    response.raise_for_status()
    
    return DeviceCode.from_dict(response.json())


def poll_for_token(device_code: DeviceCode, timeout: Optional[int] = None) -> Token:
    """
    Poll GitHub for the access token after user authorization.
    
    Args:
        device_code: DeviceCode obtained from request_device_code()
        timeout: Maximum time to poll in seconds (defaults to device_code.expires_in)
        
    Returns:
        Token containing access and refresh tokens
        
    Raises:
        TimeoutError: If authorization times out
        requests.RequestException: If the request fails
        NotAvailableError: If Copilot is not available
    """
    interval = max(device_code.interval, 5)
    deadline = time.time() + (timeout or device_code.expires_in)
    
    while time.time() < deadline:
        try:
            token = _try_get_token(device_code.device_code)
            if token:
                return token
        except _PendingError:
            time.sleep(interval)
            continue
        except _SlowDownError:
            interval += 5
            time.sleep(interval)
            continue
    
    raise TimeoutError("Authorization timed out")


class _PendingError(Exception):
    """Authorization is pending."""
    pass


class _SlowDownError(Exception):
    """Polling too fast, slow down."""
    pass


def _try_get_token(device_code: str) -> Optional[Token]:
    """
    Try to get the access token from GitHub.
    
    Args:
        device_code: The device code for polling
        
    Returns:
        Token if authorization is complete, None otherwise
        
    Raises:
        _PendingError: If authorization is still pending
        _SlowDownError: If we need to slow down polling
        NotAvailableError: If Copilot is not available
    """
    data = {
        "client_id": CLIENT_ID,
        "device_code": device_code,
        "grant_type": "urn:ietf:params:oauth:grant-type:device_code",
    }
    
    headers = {
        "Accept": "application/json",
        "Content-Type": "application/x-www-form-urlencoded",
        "User-Agent": get_headers()["User-Agent"],
    }
    
    response = requests.post(
        ACCESS_TOKEN_URL,
        data=urlencode(data),
        headers=headers,
        timeout=30,
    )
    
    result = response.json()
    error = result.get("error", "")
    
    if error == "authorization_pending":
        raise _PendingError()
    elif error == "slow_down":
        raise _SlowDownError()
    elif error:
        raise Exception(f"Authorization failed: {error}")
    
    access_token = result.get("access_token", "")
    if not access_token:
        raise _PendingError()
    
    return _get_copilot_token(access_token)


def _get_copilot_token(github_token: str) -> Token:
    """
    Exchange GitHub token for Copilot token.
    
    Args:
        github_token: GitHub OAuth access token
        
    Returns:
        Copilot token
        
    Raises:
        NotAvailableError: If Copilot is not available
        requests.RequestException: If the request fails
    """
    headers = {
        "Accept": "application/json",
        "Authorization": f"Bearer {github_token}",
        **get_headers(),
    }
    
    response = requests.get(
        COPILOT_TOKEN_URL,
        headers=headers,
        timeout=30,
        stream=True,  # Enable streaming to limit response size
    )
    
    if response.status_code == 403:
        raise NotAvailableError("GitHub Copilot is not available for this account")
    
    response.raise_for_status()
    
    # Read with size limit to prevent memory exhaustion
    content = response.raw.read(MAX_RESPONSE_SIZE + 1)
    if len(content) > MAX_RESPONSE_SIZE:
        raise ValueError("Response size exceeds maximum limit")
    
    result = json.loads(content)
    
    token = Token(
        access_token=result["token"],
        refresh_token=github_token,
        expires_at=result["expires_at"],
    )
    token.set_expires_in()
    
    return token


def refresh_token(github_token: str) -> Token:
    """
    Refresh the Copilot token using the GitHub token.
    
    Args:
        github_token: GitHub OAuth token (stored as refresh_token)
        
    Returns:
        New Copilot token
        
    Raises:
        NotAvailableError: If Copilot is not available
        requests.RequestException: If the request fails
    """
    return _get_copilot_token(github_token)

File: ts/src/copilot/oauth.ts
/**
 * GitHub Copilot OAuth device flow implementation.
 * 
 * Implements the OAuth2 device authorization flow for GitHub Copilot.
 */

import axios, { AxiosError } from "axios";
import { Token } from "../token";
import { getHeaders } from "./http";

// GitHub OAuth endpoints
const CLIENT_ID = "Iv1.b507a08c87ecfe98";
const DEVICE_CODE_URL = "https://github.com/login/device/code";
const ACCESS_TOKEN_URL = "https://github.com/login/oauth/access_token";
const COPILOT_TOKEN_URL = "https://api.github.com/copilot_internal/v2/token";

// Maximum response size (1MB) to prevent memory exhaustion
const MAX_RESPONSE_SIZE = 1024 * 1024;

/**
 * Custom error thrown when GitHub Copilot is not available.
 */
export class NotAvailableError extends Error {
  constructor(message: string = "GitHub Copilot is not available") {
    super(message);
    this.name = "NotAvailableError";
  }
}

/**
 * Device code response from GitHub.
 */
export interface DeviceCode {
  deviceCode: string;
  userCode: string;
  verificationUri: string;
  expiresIn: number;
  interval: number;
}

/**
 * Initiate the device code flow with GitHub.
 * 
 * @param timeout - Request timeout in milliseconds
 * @returns Device code information
 */
export async function requestDeviceCode(timeout: number = 30000): Promise<DeviceCode> {
  const params = new URLSearchParams({
    client_id: CLIENT_ID,
    scope: "read:user",
  });

  const headers = {
    Accept: "application/json",
    "Content-Type": "application/x-www-form-urlencoded",
    "User-Agent": getHeaders()["User-Agent"],
  };

  const response = await axios.post(DEVICE_CODE_URL, params.toString(), {
    headers,
    timeout,
  });

  return {
    deviceCode: response.data.device_code,
    userCode: response.data.user_code,
    verificationUri: response.data.verification_uri,
    expiresIn: response.data.expires_in,
    interval: response.data.interval || 5,
  };
}

/**
 * Poll GitHub for the access token after user authorization.
 * 
 * @param deviceCode - Device code from requestDeviceCode()
 * @param timeout - Maximum time to poll in milliseconds (defaults to deviceCode.expiresIn)
 * @returns Copilot token
 */
export async function pollForToken(
  deviceCode: DeviceCode,
  timeout?: number
): Promise<Token> {
  // Use let to allow interval accumulation on slow_down
  let currentInterval = Math.max(deviceCode.interval, 5) * 1000;
  const deadline = Date.now() + (timeout || deviceCode.expiresIn * 1000);

  while (Date.now() < deadline) {
    try {
      const token = await tryGetToken(deviceCode.deviceCode);
      if (token) {
        return token;
      }
    } catch (error) {
      if (error instanceof PendingError) {
        await sleep(currentInterval);
        continue;
      } else if (error instanceof SlowDownError) {
        // Accumulate interval on slow_down (matches Go behavior)
        currentInterval += 5000;
        await sleep(currentInterval);
        continue;
      }
      throw error;
    }
  }

  throw new Error("Authorization timed out");
}

/**
 * Pending authorization error.
 */
class PendingError extends Error {
  constructor() {
    super("Authorization pending");
    this.name = "PendingError";
  }
}

/**
 * Slow down polling error.
 */
class SlowDownError extends Error {
  constructor() {
    super("Slow down");
    this.name = "SlowDownError";
  }
}

/**
 * Try to get the access token from GitHub.
 */
async function tryGetToken(deviceCode: string): Promise<Token | null> {
  const params = new URLSearchParams({
    client_id: CLIENT_ID,
    device_code: deviceCode,
    grant_type: "urn:ietf:params:oauth:grant-type:device_code",
  });

  const headers = {
    Accept: "application/json",
    "Content-Type": "application/x-www-form-urlencoded",
    "User-Agent": getHeaders()["User-Agent"],
  };

  const response = await axios.post(ACCESS_TOKEN_URL, params.toString(), {
    headers,
    timeout: 30000,
  });

  const result = response.data;
  const error = result.error || "";

  if (error === "authorization_pending") {
    throw new PendingError();
  } else if (error === "slow_down") {
    throw new SlowDownError();
  } else if (error) {
    throw new Error(`Authorization failed: ${error}`);
  }

  const accessToken = result.access_token || "";
  if (!accessToken) {
    throw new PendingError();
  }

  return getCopilotToken(accessToken);
}

/**
 * Exchange GitHub token for Copilot token.
 */
async function getCopilotToken(githubToken: string): Promise<Token> {
  const headers = {
    Accept: "application/json",
    Authorization: `Bearer ${githubToken}`,
    ...getHeaders(),
  };

  try {
    const response = await axios.get(COPILOT_TOKEN_URL, {
      headers,
      timeout: 30000,
      maxContentLength: MAX_RESPONSE_SIZE,
      maxBodyLength: MAX_RESPONSE_SIZE,
    });

    const token = new Token(
      response.data.token,
      githubToken,
      0,
      response.data.expires_at
    );
    token.setExpiresIn();

    return token;
  } catch (err) {
    const axiosErr = err as AxiosError;
    if (axiosErr.isAxiosError && axiosErr.response?.status === 403) {
      throw new NotAvailableError("GitHub Copilot is not available for this account");
    }
    throw err;
  }
}

/**
 * Refresh the Copilot token using the GitHub token.
 * 
 * @param githubToken - GitHub OAuth token (stored as refresh_token)
 * @returns New Copilot token
 */
export async function refreshToken(githubToken: string): Promise<Token> {
  return getCopilotToken(githubToken);
}

/**
 * Sleep for specified milliseconds.
 */
function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}
